src/sentry/hybridcloud/models/outbox.py

```diff
@@ -224,60 +227,206 @@ def process_coalesced(
+                for i in range(0, len(all_ids), batch_size):
+                    batch_ids = all_ids[i : i + batch_size]
+                    # Use a new atomic block for each batch to ensure short and isolated transactions.
+                    with transaction.atomic(using=using):
+                        # Lock the batch of rows to ensure consistency before deletion.
+                        batch_qs = self.objects.filter(id__in=batch_ids).select_for_update(
+                            nowait=True
+                        )
+                        if batch_qs.exists():
+                            # Delete the batch and capture the count of deleted rows.
+                            batch_count = batch_qs.delete()[0]
+                            deleted_count += batch_count
+                            metrics.incr("outbox.batch_delete_success", batch_count, tags=tags)
+
+                # Finally, process the representative message (highest id) if it should not be skipped.
+                if not self.should_skip_shard():
+                    with transaction.atomic(using=using):
+                        final_qs = self.objects.filter(id=coalesced_id).select_for_update(
+                            nowait=True
+                        )
+                        if final_qs.exists():
+                            batch_count = final_qs.delete()[0]
+                            deleted_count += batch_count
+                            metrics.incr("outbox.final_delete_success", batch_count, tags=tags)
+
+                # Record metrics indicating the total number of processed messages.
+                metrics.incr("outbox.processed", deleted_count, tags=tags)
+                # Record processing lag based on when the earliest message was scheduled.
+                metrics.timing(
+                    "outbox.processing_lag",
+                    datetime.datetime.now(tz=datetime.UTC).timestamp() - scheduled_from.timestamp(),
+                    tags=tags,
+                )
+                # Record the net processing time based on when the first message was added.
+                metrics.timing(
+                    "outbox.coalesced_net_processing_time",
+                    datetime.datetime.now(tz=datetime.UTC).timestamp() - date_added.timestamp(),
+                    tags=tags,
+                )
+
+            except OperationalError as e:
+                logger.info("outbox.delete_error", extra={"error": str(e)})
+                # If an error occurs during deletion (e.g., failure to obtain a lock), increment an error metric
+                # with details of the error and re-raise the exception.
+                metrics.incr(
+                    "outbox.delete_error",
+                    tags=tags,
+                )
+                raise
+
+        except OperationalError as e:
+            if "could not obtain lock" in str(e).lower():
+                metrics.incr("outbox.lock_contention")
```

Comment:
Should we also capture an error here so we can understand which outbox categories are hitting lock contention/getting stomped?